File: crates/defi-cli/src/table.rs
use comfy_table::{Cell, Color, Table, presets::UTF8_FULL_CONDENSED};
use serde_json::Value;

/// Try to render a JSON value as a human-readable table.
/// Returns None if the shape isn't recognized.
pub fn render(value: &Value) -> Option<String> {
    // Detect by presence of key fields
    if value.get("alerts").is_some() && value.get("scan_duration_ms").is_some() {
        return Some(render_scan(value));
    }
    if value.get("holders").is_some() {
        return Some(render_whales(value));
    }
    if value.get("opportunities").is_some() && value.get("total_opportunities").is_some() {
        return Some(render_compare(value));
    }
    if value.get("arb_opportunities").is_some() && value.get("rates").is_some() {
        return Some(render_yield_scan(value));
    }
    if value.get("rates").is_some() && value.get("asset").is_some() {
        return Some(render_yield(value));
    }
    if value.get("chains").is_some() && value.get("total_alerts").is_some() {
        return Some(render_scan_all(value));
    }
    if value.get("chains").is_some() && value.get("chains_scanned").is_some() {
        return Some(render_positions(value));
    }
    if value.get("bridge").is_some() && value.get("amount_out").is_some() {
        return Some(render_bridge(value));
    }
    if value.get("aggregator").is_some() && value.get("amount_out").is_some() {
        return Some(render_swap(value));
    }
    if value.get("protocols").is_some() && value.get("summary").is_some() {
        return Some(render_status(value));
    }
    if value.get("token_balances").is_some() && value.get("lending_positions").is_some() {
        return Some(render_portfolio(value));
    }
    None
}

fn render_scan(v: &Value) -> String {
    let chain = v["chain"].as_str().unwrap_or("?");
    let ms = v["scan_duration_ms"].as_u64().unwrap_or(0);
    let count = v["alert_count"].as_u64().unwrap_or(0);
    let mut out = format!("  Scan: {} ({} ms)\n", chain, ms);

    if count > 0 {
        let mut t = Table::new();
        t.load_preset(UTF8_FULL_CONDENSED);
        t.set_header(vec![
            Cell::new("Severity").fg(Color::White),
            Cell::new("Pattern"),
            Cell::new("Asset"),
            Cell::new("Oracle"),
            Cell::new("DEX"),
            Cell::new("Gap"),
        ]);
        for a in v["alerts"].as_array().unwrap_or(&vec![]) {
            let sev = a["severity"].as_str().unwrap_or("?");
            let color = match sev {
                "critical" => Color::Red,
                "high" => Color::Yellow,
                _ => Color::Cyan,
            };
            t.add_row(vec![
                Cell::new(sev.to_uppercase()).fg(color),
                Cell::new(a["pattern"].as_str().unwrap_or("?")),
                Cell::new(a["asset"].as_str().unwrap_or("?")),
                Cell::new(format_price(a.get("oracle_price"))),
                Cell::new(format_price(a.get("dex_price"))),
                Cell::new(format!(
                    "{}%",
                    a["deviation_pct"]
                        .as_f64()
                        .map(|v| format!("{:.1}", v))
                        .unwrap_or_else(|| "?".into())
                )),
            ]);
        }
        out.push_str(&t.to_string());
    } else {
        out.push_str(&format!("  {} alerts\n", count));
    }

    // Data summary
    if let Some(data) = v.get("data") {
        let o = data
            .get("oracle_prices")
            .and_then(|v| v.as_object())
            .map(|m| m.len())
            .unwrap_or(0);
        let d = data
            .get("dex_prices")
            .and_then(|v| v.as_object())
            .map(|m| m.len())
            .unwrap_or(0);
        let s = data
            .get("stablecoin_pegs")
            .and_then(|v| v.as_object())
            .map(|m| m.len())
            .unwrap_or(0);
        out.push_str(&format!(
            "\n  Data: {} oracle, {} dex, {} stablecoin prices",
            o, d, s
        ));
    }
    out
}

fn render_scan_all(v: &Value) -> String {
    let total = v["total_alerts"].as_u64().unwrap_or(0);
    let scanned = v["chains_scanned"].as_u64().unwrap_or(0);
    let ms = v["scan_duration_ms"].as_u64().unwrap_or(0);
    let mut t = Table::new();
    t.load_preset(UTF8_FULL_CONDENSED);
    t.set_header(vec!["Chain", "Alerts", "Time", "Details"]);

    for c in v["chains"].as_array().unwrap_or(&vec![]) {
        let name = c["chain"].as_str().unwrap_or("?");
        let alerts = c["alert_count"].as_u64().unwrap_or(0);
        let cms = c["scan_duration_ms"].as_u64().unwrap_or(0);
        let details: String = c["alerts"]
            .as_array()
            .unwrap_or(&vec![])
            .iter()
            .map(|a| a["asset"].as_str().unwrap_or("?").to_string())
            .collect::<Vec<_>>()
            .join(", ");
        let color = if alerts > 0 {
            Color::Yellow
        } else {
            Color::Green
        };
        t.add_row(vec![
            Cell::new(name),
            Cell::new(alerts).fg(color),
            Cell::new(format!("{}ms", cms)),
            Cell::new(if details.is_empty() {
                "clean".into()
            } else {
                details
            }),
        ]);
    }
    format!(
        "  All-chain scan: {} chains, {} alerts, {}ms\n\n{}",
        scanned, total, ms, t
    )
}

fn render_whales(v: &Value) -> String {
    let chain = v["chain"].as_str().unwrap_or("?");
    let token = v["token"].as_str().unwrap_or("?");
    let mut t = Table::new();
    t.load_preset(UTF8_FULL_CONDENSED);
    t.set_header(vec!["#", "Address", &format!("{} Balance", token)]);

    for h in v["holders"].as_array().unwrap_or(&vec![]) {
        let rank = h["rank"].as_u64().unwrap_or(0);
        let addr = h["address"].as_str().unwrap_or("?");
        let bal = h["balance"].as_f64().unwrap_or(0.0);
        let short_addr = if addr.len() > 18 {
            format!("{}...{}", &addr[..10], &addr[addr.len() - 6..])
        } else {
            addr.to_string()
        };
        t.add_row(vec![
            Cell::new(rank),
            Cell::new(short_addr),
            Cell::new(format!("{:.2}", bal)),
        ]);
    }
    format!("  {} {} Top Holders\n\n{}", chain, token, t)
}

fn render_yield(v: &Value) -> String {
    let chain = v["chain"].as_str().unwrap_or("?");
    let asset = v["asset"].as_str().unwrap_or("?");
    let mut t = Table::new();
    t.load_preset(UTF8_FULL_CONDENSED);
    t.set_header(vec!["Protocol", "Supply APY", "Borrow APY"]);

    for r in v["rates"].as_array().unwrap_or(&vec![]) {
        let supply = r["supply_apy"].as_f64().unwrap_or(0.0);
        let borrow = r["borrow_variable_apy"].as_f64().unwrap_or(0.0);
        let color = if supply > 3.0 {
            Color::Green
        } else if supply > 1.0 {
            Color::Cyan
        } else {
            Color::White
        };
        t.add_row(vec![
            Cell::new(r["protocol"].as_str().unwrap_or("?")),
            Cell::new(format!("{:.2}%", supply)).fg(color),
            Cell::new(format!("{:.2}%", borrow)),
        ]);
    }
    let best = v["best_supply"].as_str().unwrap_or("?");
    format!(
        "  {} {} Yield Comparison (best: {})\n\n{}",
        chain, asset, best, t
    )
}

fn render_positions(v: &Value) -> String {
    let addr = v["address"].as_str().unwrap_or("?");
    let total = v
        .get("total_value_usd")
        .or(v.get("summary").and_then(|s| s.get("total_value_usd")))
        .and_then(|v| v.as_f64())
        .unwrap_or(0.0);
    let ms = v["scan_duration_ms"].as_u64().unwrap_or(0);
    let scanned = v["chains_scanned"].as_u64().unwrap_or(0);

    let mut out = format!(
        "  Positions for {}...{} ({}ms, {} chains)\n  Total: ${:.2}\n\n",
        &addr[..8],
        &addr[addr.len().saturating_sub(4)..],
        ms,
        scanned,
        total
    );

    for c in v["chains"].as_array().unwrap_or(&vec![]) {
        let chain = c["chain"].as_str().unwrap_or("?");
        let ctotal = c["chain_total_usd"].as_f64().unwrap_or(0.0);
        out.push_str(&format!("  {} (${:.2})\n", chain, ctotal));

        let mut t = Table::new();
        t.load_preset(UTF8_FULL_CONDENSED);
        t.set_header(vec!["Type", "Asset/Protocol", "Value"]);

        for b in c["token_balances"].as_array().unwrap_or(&vec![]) {
            t.add_row(vec![
                Cell::new("wallet"),
                Cell::new(b["symbol"].as_str().unwrap_or("?")),
                Cell::new(format!("${:.2}", b["value_usd"].as_f64().unwrap_or(0.0)))
                    .fg(Color::Green),
            ]);
        }
        for l in c["lending_positions"].as_array().unwrap_or(&vec![]) {
            let coll = l["collateral_usd"].as_f64().unwrap_or(0.0);
            let debt = l["debt_usd"].as_f64().unwrap_or(0.0);
            t.add_row(vec![
                Cell::new("lending"),
                Cell::new(l["protocol"].as_str().unwrap_or("?")),
                Cell::new(format!("coll ${:.0} debt ${:.0}", coll, debt)).fg(Color::Cyan),
            ]);
        }
        out.push_str(&format!("{}\n", t));
    }
    out
}

fn render_swap(v: &Value) -> String {
    let from = v["from"].as_str().unwrap_or("?");
    let to = v["to"].as_str().unwrap_or("?");
    let amt_in = v["amount_in"].as_f64().unwrap_or(0.0);
    let amt_out = v["amount_out"].as_f64().unwrap_or(0.0);
    let impact = v["price_impact_pct"]
        .as_f64()
        .map(|v| format!("{:.4}%", v))
        .unwrap_or_else(|| "n/a".into());
    let agg = v["aggregator"].as_str().unwrap_or("?");
    let chain = v["chain"].as_str().unwrap_or("?");

    format!(
        "  Swap on {} via {}\n\n  {} {} -> {:.6} {}\n  Price impact: {}\n",
        chain, agg, amt_in, from, amt_out, to, impact
    )
}

fn render_bridge(v: &Value) -> String {
    let from = v["from_chain"].as_str().unwrap_or("?");
    let to = v["to_chain"].as_str().unwrap_or("?");
    let token = v["token"].as_str().unwrap_or("?");
    let amt_in = v["amount_in"].as_f64().unwrap_or(0.0);
    let amt_out = v["amount_out"].as_f64().unwrap_or(0.0);
    let cost = v["total_cost_usd"].as_f64().unwrap_or(0.0);
    let time = v["estimated_time_sec"].as_u64().unwrap_or(0);
    let bridge = v["bridge"].as_str().unwrap_or("?");

    format!(
        "  Bridge {} -> {} via {}\n\n  {} {} -> {:.6} {}\n  Cost: ${:.2} | Time: {}s\n",
        from, to, bridge, amt_in, token, amt_out, token, cost, time
    )
}

fn render_status(v: &Value) -> String {
    let chain = v["chain"].as_str().unwrap_or("?");
    let total_p = v["summary"]["total_protocols"].as_u64().unwrap_or(0);
    let total_t = v["summary"]["total_tokens"].as_u64().unwrap_or(0);

    let mut t = Table::new();
    t.load_preset(UTF8_FULL_CONDENSED);
    t.set_header(vec!["Protocol", "Category", "Interface"]);

    for p in v["protocols"].as_array().unwrap_or(&vec![]) {
        t.add_row(vec![
            Cell::new(p["name"].as_str().unwrap_or("?")),
            Cell::new(p["category"].as_str().unwrap_or("?")),
            Cell::new(p["interface"].as_str().unwrap_or("?")),
        ]);
    }

    let empty = vec![];
    let tokens: Vec<&str> = v["tokens"]
        .as_array()
        .unwrap_or(&empty)
        .iter()
        .filter_map(|t| t.as_str())
        .collect();

    let mut out = format!("  {} — {} protocols", chain, total_p);
    if !tokens.is_empty() {
        out.push_str(&format!(
            ", {} tokens\n  Tokens: {}",
            total_t,
            tokens.join(", ")
        ));
    }
    format!("{}\n\n{}", out, t)
}

fn render_compare(v: &Value) -> String {
    let asset = v["asset"].as_str().unwrap_or("?");
    let ms = v["scan_duration_ms"].as_u64().unwrap_or(0);
    let total = v["total_opportunities"].as_u64().unwrap_or(0);

    let mut t = Table::new();
    t.load_preset(UTF8_FULL_CONDENSED);
    t.set_header(vec!["Type", "Asset", "APY", "Where", "Risk"]);

    let empty = vec![];
    for opp in v["opportunities"].as_array().unwrap_or(&empty) {
        let typ = opp["type"].as_str().unwrap_or("?");
        let apy = opp["apy"].as_f64().unwrap_or(0.0);
        let detail = opp["detail"].as_str().unwrap_or("?");
        let risk = opp["risk"].as_str().unwrap_or("?");
        let opp_asset = opp["asset"].as_str().unwrap_or("?");

        let type_label = match typ {
            "perp_funding" => "Perp Arb",
            "perp_rate" => "Perp Rate",
            "lending_supply" => "Lending",
            _ => typ,
        };

        let color = if apy.abs() > 20.0 {
            Color::Green
        } else if apy.abs() > 5.0 {
            Color::Cyan
        } else {
            Color::White
        };

        let risk_color = match risk {
            "high" => Color::Red,
            "medium" => Color::Yellow,
            _ => Color::Green,
        };

        t.add_row(vec![
            Cell::new(type_label),
            Cell::new(opp_asset),
            Cell::new(format!("{:.1}%", apy)).fg(color),
            Cell::new(detail),
            Cell::new(risk).fg(risk_color),
        ]);
    }

    format!(
        "  Yield Compare: {} ({} opportunities, {}ms)\n\n{}",
        asset, total, ms, t
    )
}

fn render_yield_scan(v: &Value) -> String {
    let asset = v["asset"].as_str().unwrap_or("?");
    let ms = v["scan_duration_ms"].as_u64().unwrap_or(0);
    let best = v["best_supply"].as_str().unwrap_or("?");

    let mut t = Table::new();
    t.load_preset(UTF8_FULL_CONDENSED);
    t.set_header(vec!["Chain", "Protocol", "Supply APY", "Borrow APY"]);

    for r in v["rates"].as_array().unwrap_or(&vec![]) {
        let supply = r["supply_apy"].as_f64().unwrap_or(0.0);
        let color = if supply > 3.0 {
            Color::Green
        } else if supply > 1.0 {
            Color::Cyan
        } else {
            Color::White
        };
        t.add_row(vec![
            Cell::new(r["chain"].as_str().unwrap_or("?")),
            Cell::new(r["protocol"].as_str().unwrap_or("?")),
            Cell::new(format!("{:.2}%", supply)).fg(color),
            Cell::new(format!(
                "{:.2}%",
                r["borrow_variable_apy"].as_f64().unwrap_or(0.0)
            )),
        ]);
    }

    let mut out = format!(
        "  {} Yield Scan ({}ms) — Best: {}\n\n{}",
        asset, ms, best, t
    );

    let empty_arbs = vec![];
    let arbs = v["arb_opportunities"].as_array().unwrap_or(&empty_arbs);
    if !arbs.is_empty() {
        let mut at = Table::new();
        at.load_preset(UTF8_FULL_CONDENSED);
        at.set_header(vec!["Spread", "Supply @", "Borrow @", "Type"]);
        for a in arbs {
            let spread = a["spread_pct"].as_f64().unwrap_or(0.0);
            let color = if spread > 1.0 {
                Color::Green
            } else {
                Color::Cyan
            };
            at.add_row(vec![
                Cell::new(format!("+{:.2}%", spread)).fg(color),
                Cell::new(format!(
                    "{} ({})",
                    a["supply_protocol"].as_str().unwrap_or("?"),
                    a["supply_chain"].as_str().unwrap_or("?")
                )),
                Cell::new(format!(
                    "{} ({})",
                    a["borrow_protocol"].as_str().unwrap_or("?"),
                    a["borrow_chain"].as_str().unwrap_or("?")
                )),
                Cell::new(a["strategy"].as_str().unwrap_or("?")),
            ]);
        }
        out.push_str(&format!("\n  Arb Opportunities\n\n{}", at));
    }
    out
}

fn render_portfolio(v: &Value) -> String {
    render_positions(v) // Same format
}

fn format_price(v: Option<&Value>) -> String {
    v.and_then(|v| v.as_f64())
        .map(|p| {
            if p > 1000.0 {
                format!("${:.0}", p)
            } else {
                format!("${:.4}", p)
            }
        })
        .unwrap_or_else(|| "?".into())
}

File: ts/packages/defi-cli/src/table.ts
import pc from "picocolors";

type JsonValue = string | number | boolean | null | JsonValue[] | { [key: string]: JsonValue };

/** Try to render a JSON value as a human-readable table. Returns null if shape isn't recognized. */
export function renderTable(value: JsonValue): string | null {
  if (typeof value !== "object" || value === null || Array.isArray(value)) return null;

  const v = value as Record<string, JsonValue>;

  if ("alerts" in v && "scan_duration_ms" in v) return renderScan(v);
  if ("holders" in v) return renderWhales(v);
  if ("opportunities" in v && "total_opportunities" in v) return renderCompare(v);
  if ("arb_opportunities" in v && "rates" in v) return renderYieldScan(v);
  if ("rates" in v && "asset" in v) return renderYield(v);
  if ("chains" in v && "total_alerts" in v) return renderScanAll(v);
  if ("chains" in v && "chains_scanned" in v) return renderPositions(v);
  if ("bridge" in v && "amount_out" in v) return renderBridge(v);
  if ("aggregator" in v && "amount_out" in v) return renderSwap(v);
  if ("protocols" in v && "summary" in v) return renderStatus(v);
  if ("token_balances" in v && "lending_positions" in v) return renderPortfolio(v);

  return null;
}

// ---------------------------------------------------------------------------
// Simple table helpers (no external dep)
// ---------------------------------------------------------------------------

function makeTable(headers: string[], rows: string[][]): string {
  const colWidths = headers.map((h, i) =>
    Math.max(h.length, ...rows.map((r) => (r[i] ?? "").replace(/\x1b\[[0-9;]*m/g, "").length))
  );

  const sep = "┼" + colWidths.map((w) => "─".repeat(w + 2)).join("┼") + "┼";
  const topBorder = "┌" + colWidths.map((w) => "─".repeat(w + 2)).join("┬") + "┐";
  const midBorder = "├" + colWidths.map((w) => "─".repeat(w + 2)).join("┼") + "┤";
  const botBorder = "└" + colWidths.map((w) => "─".repeat(w + 2)).join("┴") + "┘";

  function padCell(text: string, width: number): string {
    const visLen = text.replace(/\x1b\[[0-9;]*m/g, "").length;
    return text + " ".repeat(Math.max(0, width - visLen));
  }

  const headerRow = "│ " + headers.map((h, i) => padCell(pc.bold(h), colWidths[i])).join(" │ ") + " │";
  const dataRows = rows.map(
    (row) => "│ " + row.map((cell, i) => padCell(cell ?? "", colWidths[i])).join(" │ ") + " │"
  );

  return [topBorder, headerRow, midBorder, ...dataRows, botBorder].join("\n");
}

function asStr(v: JsonValue | undefined): string {
  if (v === undefined || v === null) return "?";
  return String(v);
}

function asF64(v: JsonValue | undefined): number {
  if (typeof v === "number") return v;
  if (typeof v === "string") return parseFloat(v) || 0;
  return 0;
}

function asU64(v: JsonValue | undefined): number {
  if (typeof v === "number") return Math.floor(v);
  if (typeof v === "string") return parseInt(v, 10) || 0;
  return 0;
}

function asArr(v: JsonValue | undefined): JsonValue[] {
  if (Array.isArray(v)) return v;
  return [];
}

function asObj(v: JsonValue | undefined): Record<string, JsonValue> {
  if (v && typeof v === "object" && !Array.isArray(v)) return v as Record<string, JsonValue>;
  return {};
}

function formatPrice(v: JsonValue | undefined): string {
  const p = asF64(v);
  if (p === 0) return "?";
  return p > 1000 ? `$${p.toFixed(0)}` : `$${p.toFixed(4)}`;
}

// ---------------------------------------------------------------------------
// Renderers
// ---------------------------------------------------------------------------

function renderScan(v: Record<string, JsonValue>): string {
  const chain = asStr(v["chain"]);
  const ms = asU64(v["scan_duration_ms"]);
  const count = asU64(v["alert_count"]);
  let out = `  Scan: ${chain} (${ms} ms)\n`;

  if (count > 0) {
    const rows = asArr(v["alerts"]).map((a) => {
      const ao = asObj(a as JsonValue);
      const sev = asStr(ao["severity"]);
      const color = sev === "critical" ? pc.red : sev === "high" ? pc.yellow : pc.cyan;
      return [
        color(sev.toUpperCase()),
        asStr(ao["pattern"]),
        asStr(ao["asset"]),
        formatPrice(ao["oracle_price"]),
        formatPrice(ao["dex_price"]),
        `${asF64(ao["deviation_pct"]).toFixed(1)}%`,
      ];
    });
    out += makeTable(["Severity", "Pattern", "Asset", "Oracle", "DEX", "Gap"], rows);
  } else {
    out += `  ${count} alerts\n`;
  }

  const data = asObj(v["data"]);
  const o = Object.keys(asObj(data["oracle_prices"])).length;
  const d = Object.keys(asObj(data["dex_prices"])).length;
  const s = Object.keys(asObj(data["stablecoin_pegs"])).length;
  out += `\n  Data: ${o} oracle, ${d} dex, ${s} stablecoin prices`;
  return out;
}

function renderScanAll(v: Record<string, JsonValue>): string {
  const total = asU64(v["total_alerts"]);
  const scanned = asU64(v["chains_scanned"]);
  const ms = asU64(v["scan_duration_ms"]);

  const rows = asArr(v["chains"]).map((c) => {
    const co = asObj(c as JsonValue);
    const alerts = asU64(co["alert_count"]);
    const cms = asU64(co["scan_duration_ms"]);
    const details = asArr(co["alerts"])
      .map((a) => asStr(asObj(a as JsonValue)["asset"]))
      .join(", ") || "clean";
    const alertStr = alerts > 0 ? pc.yellow(String(alerts)) : pc.green(String(alerts));
    return [asStr(co["chain"]), alertStr, `${cms}ms`, details];
  });

  return `  All-chain scan: ${scanned} chains, ${total} alerts, ${ms}ms\n\n` +
    makeTable(["Chain", "Alerts", "Time", "Details"], rows);
}

function renderWhales(v: Record<string, JsonValue>): string {
  const chain = asStr(v["chain"]);
  const token = asStr(v["token"]);

  const rows = asArr(v["holders"]).map((h) => {
    const ho = asObj(h as JsonValue);
    const addr = asStr(ho["address"]);
    const short = addr.length > 18 ? `${addr.slice(0, 10)}...${addr.slice(-6)}` : addr;
    return [String(asU64(ho["rank"])), short, asF64(ho["balance"]).toFixed(2)];
  });

  return `  ${chain} ${token} Top Holders\n\n` + makeTable(["#", "Address", `${token} Balance`], rows);
}

function renderYield(v: Record<string, JsonValue>): string {
  const chain = asStr(v["chain"]);
  const asset = asStr(v["asset"]);

  const rows = asArr(v["rates"]).map((r) => {
    const ro = asObj(r as JsonValue);
    const supply = asF64(ro["supply_apy"]);
    const borrow = asF64(ro["borrow_variable_apy"]);
    const color = supply > 3 ? pc.green : supply > 1 ? pc.cyan : (s: string) => s;
    return [asStr(ro["protocol"]), color(`${supply.toFixed(2)}%`), `${borrow.toFixed(2)}%`];
  });

  const best = asStr(v["best_supply"]);
  return `  ${chain} ${asset} Yield Comparison (best: ${best})\n\n` +
    makeTable(["Protocol", "Supply APY", "Borrow APY"], rows);
}

function renderPositions(v: Record<string, JsonValue>): string {
  const addr = asStr(v["address"]);
  const summary = asObj(v["summary"]);
  const total = asF64(v["total_value_usd"] ?? summary["total_value_usd"]);
  const ms = asU64(v["scan_duration_ms"]);
  const scanned = asU64(v["chains_scanned"]);

  let out = `  Positions for ${addr.slice(0, 8)}...${addr.slice(-4)} (${ms}ms, ${scanned} chains)\n  Total: $${total.toFixed(2)}\n\n`;

  for (const c of asArr(v["chains"])) {
    const co = asObj(c as JsonValue);
    const chain = asStr(co["chain"]);
    const ctotal = asF64(co["chain_total_usd"]);
    out += `  ${chain} ($${ctotal.toFixed(2)})\n`;

    const rows: string[][] = [];
    for (const b of asArr(co["token_balances"])) {
      const bo = asObj(b as JsonValue);
      rows.push(["wallet", asStr(bo["symbol"]), pc.green(`$${asF64(bo["value_usd"]).toFixed(2)}`)]);
    }
    for (const l of asArr(co["lending_positions"])) {
      const lo = asObj(l as JsonValue);
      const coll = asF64(lo["collateral_usd"]);
      const debt = asF64(lo["debt_usd"]);
      rows.push(["lending", asStr(lo["protocol"]), pc.cyan(`coll $${coll.toFixed(0)} debt $${debt.toFixed(0)}`)]);
    }
    out += makeTable(["Type", "Asset/Protocol", "Value"], rows) + "\n";
  }
  return out;
}

function renderSwap(v: Record<string, JsonValue>): string {
  const from = asStr(v["from"]);
  const to = asStr(v["to"]);
  const amtIn = asF64(v["amount_in"]);
  const amtOut = asF64(v["amount_out"]);
  const impact = typeof v["price_impact_pct"] === "number"
    ? `${asF64(v["price_impact_pct"]).toFixed(4)}%`
    : "n/a";
  const agg = asStr(v["aggregator"]);
  const chain = asStr(v["chain"]);
  return `  Swap on ${chain} via ${agg}\n\n  ${amtIn} ${from} -> ${amtOut.toFixed(6)} ${to}\n  Price impact: ${impact}\n`;
}

function renderBridge(v: Record<string, JsonValue>): string {
  const from = asStr(v["from_chain"]);
  const to = asStr(v["to_chain"]);
  const token = asStr(v["token"]);
  const amtIn = asF64(v["amount_in"]);
  const amtOut = asF64(v["amount_out"]);
  const cost = asF64(v["total_cost_usd"]);
  const time = asU64(v["estimated_time_sec"]);
  const bridge = asStr(v["bridge"]);
  return `  Bridge ${from} -> ${to} via ${bridge}\n\n  ${amtIn} ${token} -> ${amtOut.toFixed(6)} ${token}\n  Cost: $${cost.toFixed(2)} | Time: ${time}s\n`;
}

function renderStatus(v: Record<string, JsonValue>): string {
  const chain = asStr(v["chain"]);
  const summary = asObj(v["summary"]);
  const totalP = asU64(summary["total_protocols"]);
  const totalT = asU64(summary["total_tokens"]);

  const rows = asArr(v["protocols"]).map((p) => {
    const po = asObj(p as JsonValue);
    return [asStr(po["name"]), asStr(po["category"]), asStr(po["interface"])];
  });

  const tokens = asArr(v["tokens"]).map((t) => String(t)).filter(Boolean);
  let out = `  ${chain} — ${totalP} protocols`;
  if (tokens.length > 0) {
    out += `, ${totalT} tokens\n  Tokens: ${tokens.join(", ")}`;
  }
  return `${out}\n\n` + makeTable(["Protocol", "Category", "Interface"], rows);
}

function renderCompare(v: Record<string, JsonValue>): string {
  const asset = asStr(v["asset"]);
  const ms = asU64(v["scan_duration_ms"]);
  const total = asU64(v["total_opportunities"]);

  const rows = asArr(v["opportunities"]).map((opp) => {
    const oo = asObj(opp as JsonValue);
    const typ = asStr(oo["type"]);
    const apy = asF64(oo["apy"]);
    const detail = asStr(oo["detail"]);
    const risk = asStr(oo["risk"]);
    const oppAsset = asStr(oo["asset"]);

    const typeLabel = typ === "perp_funding" ? "Perp Arb"
      : typ === "perp_rate" ? "Perp Rate"
      : typ === "lending_supply" ? "Lending"
      : typ;

    const apyColor = Math.abs(apy) > 20 ? pc.green : Math.abs(apy) > 5 ? pc.cyan : (s: string) => s;
    const riskColor = risk === "high" ? pc.red : risk === "medium" ? pc.yellow : pc.green;

    return [typeLabel, oppAsset, apyColor(`${apy.toFixed(1)}%`), detail, riskColor(risk)];
  });

  return `  Yield Compare: ${asset} (${total} opportunities, ${ms}ms)\n\n` +
    makeTable(["Type", "Asset", "APY", "Where", "Risk"], rows);
}

function renderYieldScan(v: Record<string, JsonValue>): string {
  const asset = asStr(v["asset"]);
  const ms = asU64(v["scan_duration_ms"]);
  const best = asStr(v["best_supply"]);

  const rows = asArr(v["rates"]).map((r) => {
    const ro = asObj(r as JsonValue);
    const supply = asF64(ro["supply_apy"]);
    const borrow = asF64(ro["borrow_variable_apy"]);
    const color = supply > 3 ? pc.green : supply > 1 ? pc.cyan : (s: string) => s;
    return [asStr(ro["chain"]), asStr(ro["protocol"]), color(`${supply.toFixed(2)}%`), `${borrow.toFixed(2)}%`];
  });

  let out = `  ${asset} Yield Scan (${ms}ms) — Best: ${best}\n\n` +
    makeTable(["Chain", "Protocol", "Supply APY", "Borrow APY"], rows);

  const arbs = asArr(v["arb_opportunities"]);
  if (arbs.length > 0) {
    const arbRows = arbs.map((a) => {
      const ao = asObj(a as JsonValue);
      const spread = asF64(ao["spread_pct"]);
      const color = spread > 1 ? pc.green : pc.cyan;
      return [
        color(`+${spread.toFixed(2)}%`),
        `${asStr(ao["supply_protocol"])} (${asStr(ao["supply_chain"])})`,
        `${asStr(ao["borrow_protocol"])} (${asStr(ao["borrow_chain"])})`,
        asStr(ao["strategy"]),
      ];
    });
    out += "\n  Arb Opportunities\n\n" + makeTable(["Spread", "Supply @", "Borrow @", "Type"], arbRows);
  }

  return out;
}

function renderPortfolio(v: Record<string, JsonValue>): string {
  return renderPositions(v);
}
